ibis/pandas/udf.py

```diff
@@ -300,51 +327,52 @@ def execute_udf_node_groupby(op, *args, **kwargs):
         @functools.wraps(func)
         def wrapped(*args):
             return UDFNode(*args).to_expr()
-
         return wrapped
-
     return wrapper
 
 
-def udaf(input_type, output_type):
-    """Define a UDAF (user-defined aggregation function) that takes N pandas.
+def udaf(input_type, output_type, analytic=False):
+    """Define a UDAF (user-defined aggregation function) that takes N pandas
     Series or scalar values as inputs.
 
     Parameters
     ----------
-    input_type : List[T]
-        A rules encoding the abstract type of each argument or one of the types
-        found in :mod:`~ibis.expr.datatypes`. These are found in
-        :mod:`~ibis.expr.rules`. The length of this list must match the number
-        of arguments to the function. Argument splatting in the UDF signature
-        is not supported.
-    output_type : rule or ibis.expr.datatypes.DataType
-        The abstract return type of the function. This *cannot* be a rule that
-        encodes a :class:`~ibis.expr.types.ColumnExpr` since this API is for
-        defining *aggregation* functions.
+    input_type : List[ibis.expr.datatypes.DataType]
+        A list of the types found in :mod:`~ibis.expr.datatypes`. The length of
+        this list must match the number of arguments to the function. Variadic
+        arguments are not yet supported.
+    output_type : ibis.expr.datatypes.DataType
+        The return type of the function.
+    analytic : bool
+        If ``True`` this function is sensitive to group by operations.
+        Otherwise, ibis assumes this function can be called without regard to
+        grouping.
 
     Examples
     --------
     >>> import ibis
     >>> @udaf(input_type=[dt.string], output_type=dt.int64)
     ... def my_string_length_agg(series, **kwargs):
     ...     return (series.str.len() * 2).sum()
+    >>> @udaf(input_type=[dt.double], output_type=dt.double, analytic=True)
+    ... def zscore(series):  # per group, but takes N rows and return N rows
+    ...     return (series - series.mean()) / series.std()
     """
     def wrapper(func):
         UDAFNode = type(
             func.__name__,
-            (ops.Reduction,),
+            (ops.AnalyticOp if analytic else ops.Reduction,),
             {
                 'signature': sig.TypeSignature.from_dtypes(input_type),
-                'output_type': output_type.scalar_type
+                'output_type': (output_type.array_type if analytic else
+                                output_type.scalar_type)
```

Comment:
Does this mean `analytic=True` broadcasts back results but `analytic=False` doesn't?

Reply:
There's no more `analytic` flag, but to answer your question: broadcasting is only *one* of the things `analytic=True` implies. The two most important features of analytic functions are that they take N rows and return N rows and that they can be applied *per group*. I've written all of this down in the UDF documentation.